Interview WCAG

BryceStevenWilley opened the pull request:
I have been doing an accessibility audit for a docassemble interview, and these are some changes needed to be made in DA core to make the interview accessible. Possibly more to come, but there were present on most every page in the nav bar.

* added the tab role to the help/question tablist in the banner
* moved source and question tabs out of the tab list, just regular divs
  They aren't really tabs, as the ARIA definition of tabs are panels
  where only one displays at a time. Changed some CSS classes to make sure they display the same.
* hide the tab list element if empty (no help, chat, or phone). Elements with the role "tablist" are assumed to always have children, so if there aren't any children, we should remove it
* use the semantic "hidden" attribute just to make things a little clearer. (note: "aria-hidden" or "display: none" are fine for accessibility purposes)

Non interview changes:
* header tags need to start with 1 and go down, not skipping anything. I only checked the package management page and default interview pages here. Didn't change the CSS to make the headings the same size.

Used info from https://github.com/forsti0506/a11y-sitechecker and https://github.com/squizlabs/HTML_CodeSniffer (though Code Sniffer does have some false violations, thinking that elements that are actually hidden aren't).

jhpyle commented:
Note that the changes to the help buttons break the functionality. E.g, you can't switch back to the question or toggle the help tab.

BryceStevenWilley commented:
Besides for #527, I have all of the accessibility changes I'll be making for now (focused on non-developer users can interact with on a server). The only additional change (include the ones from above) was adding extra `aria-label`s to the term links, since links should generally [be understandable without surrounding context](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA8), and the linking the term name could be confusing on a screen wide list of all links.

I double checked that keyboard navigation and screen readers both work with the question and help tabs. so should be ready for another look when you get a chance @jhpyle

jhpyle commented on docassemble_base/docassemble/base/filter.py:
```diff
@@ -1467,7 +1467,11 @@ def noquote(string):
     lower_termname = re.sub(r'\s+', ' ', termname.lower(), re.DOTALL)
     if lower_termname in terms:
-        return '<a tabindex="0" class="daterm" data-bs-toggle="popover" data-bs-container="body" data-bs-placement="bottom" data-bs-content=' + noquote(markdown_to_html(terms[lower_termname]['definition'].text({}), trim=True, default_image_width='100%', do_terms=False, status=status, question=question)) + '>' + str(termname) + '</a>'
+        return '<a tabindex="0" class="daterm" aria-label=' + noquote(str(termname) + ' (term definition)') +\
```
```suggestion
        return '<a tabindex="0" class="daterm" aria-label=' + noquote(str(termname) + ' ' + word("(term definition)")) +\
```

jhpyle commented on docassemble_base/docassemble/base/filter.py:
```diff
@@ -1478,7 +1482,11 @@ def add_terms(termname, terms, label=None, status=None, question=None):
     lower_termname = re.sub(r'\s+', ' ', termname.lower(), re.DOTALL)
     if lower_termname in terms:
-        return '<a tabindex="0" class="daterm" data-bs-toggle="popover" data-bs-container="body" data-bs-placement="bottom" data-bs-content=' + noquote(markdown_to_html(terms[lower_termname]['definition'], trim=True, default_image_width='100%', do_terms=False, status=status, question=question)) + '>' + label + '</a>'
+        return '<a tabindex="0" class="daterm" aria-label=' + noquote(label + ' (term definition)') +\
```
```suggestion
        return '<a tabindex="0" class="daterm" aria-label=' + noquote(label + ' ' + word("(term definition)")) +\
```

jhpyle commented on docassemble_webapp/docassemble/webapp/server.py:
```diff
@@ -2965,20 +2965,22 @@ def make_navbar(status, steps, show_login, chat_info, debug_mode, index_params,
-        source_button = '<li class="nav-item d-none d-md-block"><button class="btn btn-link nav-link da-no-outline" title=' + json.dumps(source_message) + ' id="dasourcetoggle" data-bs-toggle="collapse" data-bs-target="#dasource"><i class="fas fa-code"></i></button></li>'
-        source_menu_item = '<a class="dropdown-item d-block d-md-none" title=' + json.dumps(source_message) + ' id="dasourcetoggle" href="#dasource" data-bs-toggle="collapse" aria-expanded="false" aria-controls="source">' + word('Source') + '</a>'
+        source_button = '<div class="nav-item navbar-nav d-none d-md-block"><button class="btn btn-link nav-link da-no-outline" title=' + json.dumps(source_message) + ' id="dasourcetoggle" data-bs-toggle="collapse" data-bs-target="#dasource"><i class="fas fa-code"></i></button></div>'
+        source_menu_item = '<a class="dropdown-item d-block d-md-none navbar" title=' + json.dumps(source_message) + ' id="dasourcetogglemenu" href="#dasource" data-bs-toggle="collapse" aria-expanded="false" aria-controls="source">' + word('Source') + '</a>'
```
Is there any reason the `id="dasourcetogglemenu"` needs to be there? `dasourcetoggle` had an `id` because there is some JavaScript that toggles `active` on the parent element, but that isn't necessary for the menu item that small devices see, I don't think.

BryceStevenWilley replied:
I don't think it needs to be there. I changed it just to avoid two elements having the same ID which is invalid HTML, but removing it is a valid option too.

jhpyle commented on docassemble_webapp/docassemble/webapp/templates/pages/interviews.html:
```diff
@@ -5,7 +5,7 @@
 {%- extends 'flask_user/public_base.html' %}
 {%- block content %}
-            <h3 class="da-page-header dainterviewpage">{{ get_part('interview page heading', word('Resume an interview')) }}</h3>
+            <h1 class="da-page-header dainterviewpage">{{ get_part('interview page heading', word('Resume an interview')) }}</h1>
```
```suggestion
            <h1 class="da-page-header dainterviewpage h3">{{ get_part('interview page heading', word('Resume an interview')) }}</h1>
```

jhpyle commented on docassemble_webapp/docassemble/webapp/templates/pages/update_package.html:
```diff
@@ -3,6 +3,8 @@
 {% block content %}
 {% from "base_templates/form_macros.html" import render_field, render_file_field, render_select_field, render_submit_field, render_checkbox_field with context %}
 
+<h1>{{ word('Package Management') }}</h1>
```
```suggestion
```

jhpyle replied:
This just repeats what is in the navigation bar.

BryceStevenWilley replied:
That's fair. I did it because the page does need an h1, and I didn't want to change too much, but I agree with your suggestion to make "Package Management" the h1.

jhpyle commented on docassemble_webapp/docassemble/webapp/templates/pages/update_package.html:
```diff
@@ -3,6 +3,8 @@
+<h1>{{ word('Package Management') }}</h1>
+
 {%- if allowed_to_upgrade %}
 <h2>{{ word('Upgrade docassemble') }}</h2>
```
```suggestion
<h1 class="h2">{{ word('Upgrade docassemble') }}</h1>
```

jhpyle commented on docassemble_webapp/docassemble/webapp/templates/pages/update_package.html:
```diff
@@ -16,12 +18,12 @@ <h2>{{ word('Install or update a package') }}</h2>
 <form action="{{ url_for('update_package') }}" method="POST" class="form" role="form" enctype="multipart/form-data">
   {{ form.hidden_tag() }}
-  <h4>{{ word('You can provide a:') }}</h4>
+  <h3>{{ word('You can provide a:') }}</h3>
```
```suggestion
  <h2 class="h3">{{ word('You can provide a:') }}</h2>
```

BryceStevenWilley replied:
Ended up changing these back to h3, but keeping the h4 class, since they are all sub headings of "Install or update a package", and shouldn't be at the same level.

jhpyle commented on docassemble_webapp/docassemble/webapp/templates/pages/update_package.html:
```diff
@@ -16,12 +18,12 @@ <h2>{{ word('Install or update a package') }}</h2>
   {{ render_field(form.giturl, placeholder='https://github.com/username/packagename') }}
   {{ render_select_field(form.gitbranch) }}
-  <h4>{{ word('Or you can upload a:') }}</h4>
+  <h3>{{ word('Or you can upload a:') }}</h3>
```
```suggestion
  <h2 class="h3">{{ word('Or you can upload a:') }}</h2>
```

jhpyle commented on docassemble_webapp/docassemble/webapp/templates/pages/update_package.html:
```diff
@@ -16,12 +18,12 @@ <h2>{{ word('Install or update a package') }}</h2>
+  <h3>{{ word('Or you can upload a:') }}</h3>
   {{ render_file_field(form.zipfile) }}
-  <h4>{{ word('Or you can specify a:') }}</h4>
+  <h3>{{ word('Or you can specify a:') }}</h3>
```
```suggestion
  <h2 class="h4">{{ word('Or you can specify a:') }}</h2>
```

BryceStevenWilley commented:
I double checked that removing `dasourcetogglemenu` works fine (it does) and I accepted most of the suggestions (kept some sub headings of "Install or update a package" at h3, but accepted the suggestion to make the style h4).

BryceStevenWilley commented:
Did another quick change: didn't realize that emoji in terms weren't already converted by the time that the term label is added, and it was trying to add the emoji HTML in the middle of the aria-label. I fixed this by just adding "(the-emoji-name emoji)" to the aria-label.

BryceStevenWilley commented:
The last time I'll add more to this PR, but unfortunately the changes in [the recent commit about required fieldsets](https://github.com/jhpyle/docassemble/commit/8bc120ce5af0fbad02fe00252aa32d4f17be62d0) aren't [valid ARIA](https://github.com/w3c/aria/issues/1322), so I reverted them, and changed some of the descriptions.

jhpyle commented on docassemble_webapp/docassemble/webapp/server.py:
```diff
@@ -2844,7 +2844,7 @@ def progress_bar(progress, interview):
     else:
         percentage = ''
-    return '<div class="progress mt-2"><div class="progress-bar" role="progressbar" aria-valuenow="' + str(progress) + '" aria-valuemin="0" aria-valuemax="100" style="width: ' + str(progress) + '%;">' + percentage + '</div></div>\n'
+    return '<div class="progress mt-2"><div class="progress-bar" aria-label="Interview Progress" role="progressbar" aria-valuenow="' + str(progress) + '" aria-valuemin="0" aria-valuemax="100" style="width: ' + str(progress) + '%;">' + percentage + '</div></div>\n'
```
I'll fix this when I merge the PR, but note that this introduces hard-coded English. I changed it to `aria-label="' + noquote(word('Interview Progress')) + '"`

BryceStevenWilley replied:
Thanks, sorry I keep missing those!
